Library/Homebrew/cli_parser.rb

```diff
@@ -0,0 +1,271 @@
+    # Returns `given_args` and `unnamed_args`
+    # `given_args`  : A hash with option instance as key and
+    #                 their respective parameters as value for all args provided on CLI
+    # `unnamed_args`: An array containing remaining unnamed args
+    def separate_given_unnamed_args(cmdline_args)
+      given_args = {}
+      unnamed_args = each_arg cmdline_args do |arg, params|
+        argument = case arg
+        when /^--([^-]\S*)$/
+          name = convert_arg_to_sym(Regexp.last_match(1))
+          @declared_args[name]
+        else
+          raise InvalidArgError
+        end
+
+        next nil if ignore_invalid_options && !argument
+        raise UnknownArgError, arg unless argument
+
+        given_args[argument] ||= {}
+        given_args[argument][:arg] = arg
+        given_args[argument][:params] ||= []
+
+        num_params_taken = 0
+        next num_params_taken if params.empty?
+        # Support only args which take a single param for now
+        if argument.single_arg?
+          given_args[argument][:params] = params[0]
+          num_params_taken = 1
+        end
+        num_params_taken
+      end
+      [given_args, unnamed_args]
+    end
+
+    # Check if required args are among given args
+    def check_required_args(given_args)
+      @declared_args.each do |arg_name, option|
+        next unless option.required?
+        next if given_args.key? arg_name
+        raise RequiredArgError, option
+      end
+    end
+
+    def parse_given_args(given_args)
+      parsed_args = populate_default_arg_params
+      given_args.each do |opt, given_data|
+        arg, params = given_data.values_at :arg, :params
+        if params.nil? && !opt.flag?
+          raise MissingArgParamError, arg unless opt.default
+          params = opt.default
+        end
+
+        case opt.type
+        when :string
+          parsed_args[opt.name] = params.to_s
+        when :comma_array
+          parsed_args[opt.name] = params.to_s.split(",")
+        when :flag
+          parsed_args[opt.name] = true
+        end
+      end
+      parsed_args
+    end
+
+    # For each element in ARGV
+    # Yields to a block with the argument and its parameters
+    # Returns array of unnamed args
+    def each_arg(args)
+      remaining_args = []
+      i = 0
+
+      until i >= args.length
+        case args[i]
+
+        # match: --argument=<params>
+        when /^--(\S+?)=(.*)$/
+          num_params_taken = yield "--#{Regexp.last_match(1)}", [Regexp.last_match(2)]
+          remaining_args << args[i] if num_params_taken.nil?
+
+        # match: --argument param1 param2
+        when /^--(\S+)$/
+          params = collect_argument_parameters(args, i + 1)
+          num_params_taken = yield args[i], params
+          if num_params_taken.nil?
+            remaining_args << args[i]
+          else
+            i += num_params_taken
+          end
+        else
+          remaining_args << args[i]
+        end
+        i += 1
+      end
```

Comment:
I think it would be more readable if we duplicated the string and then just shifted arguments off it, rather than using an `i` variable that has to be manually incremented.